server/datastore/mysql/apple_mdm.go

```diff
@@ -875,3 +875,102 @@ VALUES
 		return nil
 	})
 }
+
+func (ds *Datastore) ListMDMAppleProfilesToInstall(ctx context.Context) ([]*fleet.MDMAppleProfilePayload, error) {
+	query := `
+          SELECT macp.profile_id, h.uuid as host_uuid
+          FROM mdm_apple_configuration_profiles macp
+          LEFT JOIN hosts h
+              ON h.team_id = macp.team_id OR (h.team_id IS NULL AND macp.team_id = 0)
+          LEFT JOIN host_mdm_apple_profiles hmap
+              ON hmap.profile_id = macp.profile_id AND hmap.host_uuid = h.uuid
+          WHERE
+	    hmap.profile_id IS NULL
+	    AND hmap.host_uuid IS NULL
+	    AND hmap.status != 'pending' OR hmap.status IS NULL
+	    AND hmap.operation_type != 'install' OR hmap.operation_type IS NULL
+	    -- accounts for the edge case of having profiles but not having hosts 
+	    AND h.uuid IS NOT NULL
+	`
+
+	var profiles []*fleet.MDMAppleProfilePayload
+	err := sqlx.SelectContext(ctx, ds.reader, &profiles, query)
+	return profiles, err
+}
+
+func (ds *Datastore) ListMDMAppleProfilesToRemove(ctx context.Context) ([]*fleet.MDMAppleProfilePayload, error) {
+	query := `
+          SELECT hmap.profile_id, hmap.host_uuid, macp.identifier as profile_identifier
+          FROM mdm_apple_configuration_profiles macp
+          LEFT JOIN hosts h
+              ON h.team_id = macp.team_id OR (h.team_id IS NULL AND macp.team_id = 0)
+          RIGHT JOIN host_mdm_apple_profiles hmap
+              ON hmap.profile_id = macp.profile_id AND hmap.host_uuid = h.uuid
+          WHERE
+	      macp.profile_id IS NULL
+	      AND h.uuid IS NULL
+	      AND hmap.status != 'pending' OR hmap.status IS NULL
+	      AND hmap.operation_type != 'remove'
+	`
```

Comment:
@mna @gillespi314 To send the `RemoveProfile` command to the host, we need the `PayloadIdentifier` of the profile.

Seems like we either need to add a `profile_identifier` column to `host_mdm_apple_profiles` or soft-delete `mdm_apple_configuration_profiles`, any thoughts/ideas?

Reply:
I think it'd be reasonable to add the `profile_identifier` to `host_mdm_apple_profiles`, given that the profile identifier cannot change for a given `profile_id`, so the cost of maintaining this extra column should be low (zero? Like we would never have to update it, I think?). If that helps simplify things, of course.

To be honest, I can't parse that query and really understand how it works, the LEFT JOIN and RIGHT JOIN that uses both the main table and the left-joined table in the ON clause, I'm not really sure how that works :D 

Haven't looked at the rest of the PR, so that may be why I struggle to understand it in isolation, but assuming this is to get the list of profiles to remove from hosts, shouldn't the join with hosts be a standard join? I.e. the host needs to exist, otherwise we don't have to remove the profile from it?

Reply:
> I think it'd be reasonable to add the profile_identifier to host_mdm_apple_profiles, given that the profile identifier cannot change for a given profile_id, so the cost of maintaining this extra column should be low (zero? Like we would never have to update it, I think?). If that helps simplify things, of course

thanks!

> To be honest, I can't parse that query and really understand how it works, the LEFT JOIN and RIGHT JOIN that uses both the main table and the left-joined table in the ON clause, I'm not really sure how that works :D

would writing the query like this make it more parseable? I wish we could use CTEs

```sql
	      SELECT hmap.profile_id, hmap.host_uuid
	      FROM (
	      	SELECT h.uuid, macp.profile_id FROM mdm_apple_configuration_profiles macp
          	JOIN hosts h
              ON h.team_id = macp.team_id OR (h.team_id IS NULL AND macp.team_id = 0)
              WHERE macp.profile_id IS NULL AND h.uuid IS NULL
          ) as desired_state
          RIGHT JOIN host_mdm_apple_profiles hmap
              ON hmap.profile_id = desired_state.profile_id AND hmap.host_uuid = desired_state.uuid
	      AND hmap.status != 'pending' OR hmap.status IS NULL
	      AND hmap.operation_type != 'remove'
```

> Haven't looked at the rest of the PR, so that may be why I struggle to understand it in isolation, but assuming this is to get the list of profiles to remove from hosts, shouldn't the join with hosts be a standard join? I.e. the host needs to exist, otherwise we don't have to remove the profile from it?

I think you're right, thanks!!

Reply:
It looks like you're selecting `NULL, NULL` from the `desired_state` query?

```
SELECT 
      h.uuid, 
      macp.profile_id 
    FROM 
     ...
    WHERE 
      macp.profile_id IS NULL 
      AND h.uuid IS NULL
```

That doesn't seem right or am I missing something?

Reply:
@mna so sorry, I wrote that without testing, I meant:

```sql
SELECT *
  FROM (
  	SELECT h.uuid, macp.profile_id
  	FROM mdm_apple_configuration_profiles macp 
  	JOIN hosts h
  	ON h.team_id = macp.team_id OR (h.team_id IS NULL AND macp.team_id = 0)
  ) as desired_state
  RIGHT JOIN host_mdm_apple_profiles hmap
      ON hmap.profile_id = desired_state.profile_id AND hmap.host_uuid = desired_state.uuid
  WHERE desired_state.profile_id IS NULL AND desired_state.uuid IS NULL
  AND hmap.status != 'pending' OR hmap.status IS NULL
  AND hmap.operation_type != 'remove'
```

What I'm trying to represent, if:

- Set `A` which is the desired state, can be obtained from a `JOIN` between `mdm_apple_configuration_profiles` and `hosts`
- Set `B` is the current state given by `host_mdm_apple_profiles`

Then:

- `A - B` gives the profiles to be installed
- `B - A` gives the profiles to be removed

But maybe there's an easier way to think about this?

Reply:
Thanks for the explanation! That set A-B thing is very helpful, might be worth adding as a comment next to the relevant queries for "future us". The reasoning makes total sense to me! One little nit, any reason you store the host's `uuid` in `host_mdm_apple_profiles` instead of the `host_id`? It would probably speed up the lookups a bit while making the table take less space (unless of course the `uuid` is required for apple to execute the commands, although even then you could get it from the join with `hosts`).

Reply:
This makes sense to me too.

> I wish we could use CTEs
+1